jpa/hibernate5_3/src/main/java/org/jboss/as/jpa/hibernate5/HibernatePersistenceProviderAdaptor.java

```diff
@@ -99,32 +99,62 @@ public void addProviderDependencies(PersistenceUnitMetadata pu) {
+        if (Classification.NONE.equals(platform.defaultCacheClassification())) {
             JPA_LOGGER.tracef("second level cache is not supported in platform, ignoring shared cache mode");
             pu.setSharedCacheMode(SharedCacheMode.NONE);
         }
-        // check if 2lc is explicitly disabled which takes precedence over other settings
-        boolean sharedCacheDisabled = SharedCacheMode.NONE.equals(pu.getSharedCacheMode())
-                ||
-                NONE.equals(sharedCacheMode) ||
-                SharedCacheMode.UNSPECIFIED.equals(pu.getSharedCacheMode()) ||
-                UNSPECIFIED.equals(sharedCacheMode);
-        if (!sharedCacheDisabled &&
-                (null == properties.getProperty(AvailableSettings.USE_SECOND_LEVEL_CACHE) ||
-                        Boolean.parseBoolean(properties.getProperty(AvailableSettings.USE_SECOND_LEVEL_CACHE))))
-                {
+        // precedence order of cache settings (1 overrides other settings and 3 is lowest precedence level):
+        // 1 - SharedCacheMode.NONE
+        // 2 - AvailableSettings.USE_SECOND_LEVEL_CACHE
+        // 2 - AvailableSettings.USE_QUERY_CACHE
+        // 3 - SharedCacheMode.UNSPECIFIED =
+        // 3 - SharedCacheMode.ENABLE_SELECTIVE
+        // 3 - SharedCacheMode.DISABLE_SELECTIVE
+
+        // if SharedCacheMode.NONE, disable cache.
+        // if Hibernate cache settings are set to false, disable cache
+        boolean cacheDisabled = noneCacheMode(pu) || (haveHibernateCacheProperty(pu) && !hibernateCacheEnabled(pu));
+
+        if (!cacheDisabled) {
             HibernateSecondLevelCache.addSecondLevelCacheDependencies(pu.getProperties(), pu.getScopedPersistenceUnitName());
             JPA_LOGGER.tracef("second level cache enabled for %s", pu.getScopedPersistenceUnitName());
+            // for SharedCacheMode.UNSPECIFIED, enable the cache and enable caching for entities marked with Cacheable
+            if (unspecifiedCacheMode(pu)) {
+                pu.setSharedCacheMode(SharedCacheMode.ENABLE_SELECTIVE);
+            }
+
         } else {
             JPA_LOGGER.tracef("second level cache disabled for %s, pu %s property = %s, pu.getSharedCacheMode = %s",
                     pu.getScopedPersistenceUnitName(),
                     SHARED_CACHE_MODE,
                     sharedCacheMode,
                     pu.getSharedCacheMode().toString());
-            pu.setSharedCacheMode(SharedCacheMode.NONE);  // ensure that Hibernate doesn't try to use the 2lc for UNSPECIFIED
+            pu.setSharedCacheMode(SharedCacheMode.NONE);  // ensure that Hibernate doesn't try to use the 2lc
         }
     }
 
+    private boolean haveHibernateCacheProperty(PersistenceUnitMetadata pu) {
+        return (pu.getProperties().getProperty(AvailableSettings.USE_SECOND_LEVEL_CACHE) != null ||
+                pu.getProperties().getProperty(AvailableSettings.USE_QUERY_CACHE) != null);
+    }
```

Comment:
I don't think this method is need as hibernateCacheEnabled has to deal with null anyway (and does) and it would return false if both props are null, the same as this method does.

Reply:
hibernateCacheEnabled is only returning true for cache enabled and false for cache disabled, but we need a 3rd state to know that the application didn't specify the USE_SECOND_LEVEL_CACHE/USE_QUERY_CACHE properties.

I'll update the comment (if Hibernate cache settings are set to false, disable cache), to make this clearer.